cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -549,6 +549,20 @@ hash_value_type CUDA_DEVICE_CALLABLE MurmurHash3_32<double>::operator()(double c
   return this->compute_floating_point(key);
 }
 
+template <>
+hash_value_type CUDA_DEVICE_CALLABLE
+MurmurHash3_32<numeric::decimal32>::operator()(numeric::decimal32 const& key) const
+{
+  return this->compute(key.value());
+}
+
+template <>
+hash_value_type CUDA_DEVICE_CALLABLE
+MurmurHash3_32<numeric::decimal64>::operator()(numeric::decimal64 const& key) const
```

Comment:
if we're dispatching storage type, should these specializations even need to exist?

Reply:
Good question. It's kind of the opposite. The specialization in `hash_functions.cuh` takes care of the `serail_murmur_hash3_32` which is used for `HASH_SERIAL_MURMUR3` and `HASH_SPARK_MURMUR3` algorithms. The `SparkMurmurHash3_32` specialization for `decimal32` and `decimal64` cast the value component to `uint64_t`. I'm told this is what Spark expects -- using `type_dispatch<dispatch_storage_type>` here will make this fail since the neither storage-type for `decimalXX` is `uint64_t`.

So this means only the specialization for `MurmurHash3_32` is actually required here to fix the issue. The change to the `row_hasher` to use `type_dispatch<dispatch_storage_type>` was just a optimization to generate less code for the non-serial hash algorithms (e.g. `HASH_MURMUR3` and `HASH_MD5`) as well as other places the `row_hasher` may be used.

Reply:
Thank you for this explanation, @davidwendt.